Support TOML lists in cmake.define

alexreinking opened the pull request:
Adds support for specifying CMake lists in `cmake.define` using TOML lists:

```
[tool.scikit-build.cmake.define]
ONE_LEVEL_LIST = [
    "Foo",
    "Bar",
    "ExceptionallyLargeListEntryThatWouldOverflowTheLine",
    "Baz",
]
NESTED_LIST = [ "Apple", "Lemon;Lime", "Banana" ]
```

Escapes semicolons in list items so as to allow nested lists.

Fixes #874

LecrisUT commented on src/scikit_build_core/settings/skbuild_model.py:
```diff
@@ -27,6 +27,23 @@ def __dir__() -> List[str]:
 
 
+class CMakeSettingsDefine(str):
+    json_schema = Union[str, bool, List[str]]
```
Iirc the CI fails because this is not in `__all__`. Not sure if the intention is to be there or not. But at the very least can you add a quick docstring?

alexreinking replied:
Added a docstring and marked with `_` prefix to indicate it's a private type.

LecrisUT replied:
That didn't work, iirc the issue is that `CMakeSettingsDefine` is used in the documentation of `CMakeSettings` and it fails when it's trying to reference it.

I wonder if you could also make it a standard class with `__str__` implementation. It would at least be less confusing compared to having the type change within the constructor of `CMakeSettings`

alexreinking replied:
I tried it as a standard class with a `__str__` implementation, but it grew the diff by a good bit. Trying to define it as a dataclass really got the schema code confused, too.

LecrisUT replied:
Yeah, definitely not a dataclass in this case. Big diffs are not necessarily a problem as long as they are easy to follow. For this one I'm ok with either implementation, but let's see what Henry preferrs as well.

henryiii replied:
Yep, can't be a Dataclass. I think this is fine - where is the type changing? CMakeSettings doesn't have a constructor, it's a Dataclass, and this is how you communicate to it that you want something special?

henryiii replied:
This `__new__` does return a `CMakeDefine`, though? I don’t see how a `str` could be returned.

LecrisUT replied:
Ah, right, I misread the `super.__new__`. All cool in this case then I guess.

LecrisUT commented on tests/packages/cmake_defines/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+cmake_minimum_required(VERSION 3.26)
```
```suggestion
cmake_minimum_required(VERSION 3.15)
```
Should add a badge for the minimum CMake

alexreinking replied:
Done, thanks!

LecrisUT commented:
Can you add some docs on the change, particularly regarding the nested list syntax?
https://github.com/scikit-build/scikit-build-core/blob/7636e7bd59fc7ac671905eadf3ad605bebba8a87/docs/configuration.md?plain=1#L408

Maybe as a `:::{versionchanged} 0.11` inside the tab block:
https://github.com/scikit-build/scikit-build-core/blob/7636e7bd59fc7ac671905eadf3ad605bebba8a87/docs/configuration.md?plain=1#L418-L425

henryiii reviewed (approved):
A mention in the docs, and one unit test would be nice, but not required. Thanks! Not sure I would have come up with this solution.

alexreinking commented:
@henryiii - I added a unit test to `test_skbuild_settings.py` that checks the lists in `packages/cmake_defines/pyproject.toml` are converted to the correct strings. I also updated the documentation. This is how I see it rendered locally:

<img width="839" alt="image" src="https://github.com/user-attachments/assets/c8b260a1-f34f-4544-964f-676d632a40a6">

henryiii commented:
I've adjusted the docs style to match the existing style a bit better, otherwise looks fantastic.

alexreinking commented:
> I've adjusted the docs style to match the existing style a bit better, otherwise looks fantastic.

Thanks! Happy to contribute 🙂

alexreinking commented:
@henryiii -- not sure why one of the CI bots failed. Hoping it's just an intermittent failure. Merged in main to re-run.

henryiii commented:
Yes, it is. Even with a rebuild PyPy is flaky in pretty much every repo.
